File: app/src/main/java/com/relic/presentation/displaypost/DisplayPostVM.java
package com.relic.presentation.displaypost;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MediatorLiveData;
import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModel;
import android.support.annotation.Nullable;
import android.util.Log;

import com.relic.data.CommentRepository;
import com.relic.data.ListingRepository;
import com.relic.data.PostRepository;
import com.relic.data.gateway.PostGateway;
import com.relic.data.models.CommentModel;
import com.relic.data.models.PostModel;
import com.relic.presentation.callbacks.RetrieveNextListingCallback;

import java.util.List;

public class DisplayPostVM extends ViewModel implements DisplayPostContract.ViewModel {
  private String TAG = "DISPLAYPOST_VM";

  private ListingRepository listingRepo;
  private PostRepository postRepo;
  private CommentRepository commentRepo;
  private PostGateway postGateway;

  private MediatorLiveData<PostModel> currentPost;
  private MediatorLiveData<List<CommentModel>> commentList;
  private MediatorLiveData <String> commentListingKey;

  private String postFullname;
  private String subName;
  private boolean initialized;
  private boolean alreadyEmpty;

  public void init(ListingRepository listingRepo, PostRepository postRepo, CommentRepository commentRepo, String subreddit, String fullname) {
    if (!initialized) {
      // initialize reference to repos for this VM
      this.listingRepo = listingRepo;
      this.postRepo = postRepo;
      this.commentRepo = commentRepo;
      postGateway = postRepo.getPostGateway();

      subName = subreddit;
      postFullname = fullname;

      currentPost = new MediatorLiveData<>();
      commentList = new MediatorLiveData<>();
      commentListingKey = new MediatorLiveData<>();

      observeLivedata();
      retrieveMoreComments(true);
      initialized = true;
    }
  }


  /**
   * Exposes the post to the view
   * @return post as livedata
   */
  public LiveData<PostModel> getPost() {
    return currentPost;
  }


  /**
   * Exposes the list of comments to the view
   * @return comment list as livedata
   */
  public LiveData<List<CommentModel>> getCommentList() {
    return commentList;
  }


  /**
   * Add sources and listeners to all local livedata
   */
  private void observeLivedata() {
    // retrieves the livedata post to be exposed to the view
    currentPost.addSource(postRepo.getPost(postFullname),
        (PostModel post) -> {
          if (post != null ) {
            // TODO add additional actions to trigger when post loaded
            currentPost.setValue(post);
          }
        });

    // retrieve the comment list as livedata so that we can expose it to the view first
    commentList.addSource(commentRepo.getComments(postFullname),
        (List<CommentModel> comments) -> {
          if (comments.isEmpty()) {
            // retrieve more comments if we detect that none are stored locally
            commentRepo.retrieveComments(subName, postFullname, null);
          }
          else {
            // TODO add additional actions to trigger when comments loaded
            commentList.setValue(comments);
          }
        });

    commentListingKey.addSource(listingRepo.getKey(),
        (String nextListingKey) -> {
          if (nextListingKey != null) {
            // retrieve the next listing using its key
            commentRepo.retrieveComments(subName, postFullname, nextListingKey);
          }
        });

  }

  /**
   * Refreshes the post and comment data from network
   */
  public void refresh() {
    // retrieves post from network
    postRepo.retrievePost(subName, postFullname);
    // retrieves comments form network
    retrieveMoreComments(true);
  }


  @Override
  public void retrieveMoreComments(boolean refresh) {
    if (refresh) {
      // delete comments stored locally, let observer retrieve more once it registers the change
      commentRepo.clearComments(postFullname);
    }
    else {
      // retrieve the next listing for the comments on this post
      listingRepo.retrieveKey(postFullname);
    }
  }

}

File: app/src/main/java/com/relic/presentation/displaypost/DisplayPostVM.kt
package com.relic.presentation.displaypost

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.MutableLiveData
import com.relic.data.CommentRepository
import com.relic.data.ListingRepository
import com.relic.data.PostRepository
import com.relic.data.gateway.PostGateway
import com.relic.data.repository.NetworkException
import com.relic.domain.models.CommentModel
import com.relic.domain.models.PostModel
import com.relic.network.NetworkUtil
import com.relic.network.request.RelicRequestError
import com.relic.presentation.base.RelicViewModel
import com.shopify.livedataktx.SingleLiveData
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext

class DisplayPostVM (
    private val postRepo : PostRepository,
    private val commentRepo: CommentRepository,
    private val listingRepo: ListingRepository,
    private val postGateway: PostGateway,
    private val networkUtil : NetworkUtil,
    private val subName: String,
    private val postFullname: String
): RelicViewModel(), DisplayPostContract.ViewModel {

    class Factory @Inject constructor(
        private val postRepo: PostRepository,
        private val commentRepo: CommentRepository,
        private val listingRepo: ListingRepository,
        private val networkUtil : NetworkUtil,
        private val postGateway: PostGateway
    ) {
        fun create(
            subredditName : String,
            postFullname : String
        ) : DisplayPostVM {
            return DisplayPostVM(postRepo, commentRepo, listingRepo, postGateway, networkUtil, subredditName, postFullname)
        }
    }

    private val _postLiveData = MediatorLiveData<PostModel> ()
    private val _commentListLiveData = MediatorLiveData<List<CommentModel>> ()
    private val _errorLiveData = MutableLiveData<PostErrorData> ()

    val postLiveData : LiveData<PostModel> =_postLiveData
    val commentListLiveData : LiveData<List<CommentModel>> = _commentListLiveData
    val errorLiveData : LiveData<PostErrorData> = _errorLiveData

    init {
        _postLiveData.addSource<PostModel>(postRepo.getPost(postFullname)) { post ->
            _postLiveData.postValue(post)
        }

        if (networkUtil.checkConnection()) {
            refreshData()
        } else {
            launch(Dispatchers.IO) {
                val localComments = commentRepo.getComments(postFullname)
                Timber.d(localComments.size.toString())
                _commentListLiveData.postValue(localComments)
            }

            publishException(PostErrorData.NetworkUnavailable)
        }
    }

    override fun refreshData() {
        if (networkUtil.checkConnection()) {
            launch(Dispatchers.Main) {
                val commentsAndPost = commentRepo.retrieveComments(subName, postFullname, refresh = true)


                // TODO when adding the preferences manager, check if user wants to save comments
                launch {
                    // remove previous comments for this post and stores new results
                    commentRepo.deleteComments(postFullname)
                    commentRepo.insertComments(commentsAndPost.comments)
                }

                _postLiveData.postValue(commentsAndPost.post)
                _commentListLiveData.postValue(commentsAndPost.comments)
            }
        }
        else {
            publishException(PostErrorData.NetworkUnavailable)
        }
    }

    private fun removeReplies (position : Int) {
        _commentListLiveData.value?.toMutableList()?.let { commentList ->
            val parentDepth = commentList[position].depth
            var itemsToRemove = 0

            // hide anything with depth > item.depth until reaching another item of the same depth
            var comparePosition = position + 1
            while (comparePosition < commentList.size && commentList[comparePosition].depth > parentDepth) {
                // add up number of items to be removed
                itemsToRemove ++
                comparePosition ++
            }

            commentList.subList(position + 1, position + 1 + itemsToRemove).clear()
            _commentListLiveData.postValue(commentList)
        }
    }

    /**
     * converts exceptions we handle into exceptions we've defined in the contract and posts
     * it to the livedata to let the view display it
     */
    private fun publishException(exception : Exception) {
        val viewException : PostErrorData = when (exception) {
            is RelicRequestError -> PostErrorData.NetworkUnavailable
            else -> PostErrorData.UnexpectedException
        }

        publishException(viewException)
    }

    private fun publishException(viewException : PostErrorData) {
        _errorLiveData.postValue(viewException)
    }

    //  region view action delegate

    override fun onExpandReplies(comment: CommentModel, expanded : Boolean) {
        launch(Dispatchers.Main){
            val deferredParentPos = async { _commentListLiveData.value!!.indexOfFirst { it.fullName == comment.fullName } }

            if (expanded) {
                removeReplies(deferredParentPos.await())
            } else {

                val moreReplies = commentRepo.retrieveCommentChildren(postFullname, comment)

                _commentListLiveData.value?.let { commentList ->
                    val newList = commentList.toMutableList().apply {
                        val parentPos = deferredParentPos.await()
                        removeAt(parentPos)
                        addAll(parentPos, moreReplies)
                    }

                    _commentListLiveData.postValue(newList)
                }
            }
        }
    }

    // endregion view action delegate

    override fun handleException(context: CoroutineContext, e: Throwable) {
        val postE = when (e) {
            is NetworkException -> PostErrorData.NetworkUnavailable
            else -> PostErrorData.UnexpectedException
        }

        publishException(postE)
    }
}
